File: src/main/java/frc/robot/subsystems/intake/IntakeIOReal.java
package frc.robot.subsystems.intake;

import com.ctre.phoenix6.hardware.CANcoder;
import com.revrobotics.PersistMode;
import com.revrobotics.REVLibError;
import com.revrobotics.RelativeEncoder;
import com.revrobotics.ResetMode;
import com.revrobotics.spark.SparkLowLevel.MotorType;
import com.revrobotics.spark.config.SparkBaseConfig.IdleMode;
import com.revrobotics.spark.config.SparkMaxConfig;

import edu.wpi.first.units.Units;
import edu.wpi.first.units.measure.Angle;
import edu.wpi.first.units.measure.AngularVelocity;
import edu.wpi.first.units.measure.Voltage;
import edu.wpi.first.wpilibj.Alert;
import edu.wpi.first.wpilibj.Alert.AlertType;
import frc.utils.controlWrappers.ArmFF;
import frc.utils.controlWrappers.ProfiledPID;
import frc.utils.controlWrappers.SimpleFF;
import frc.utils.motorWrappers.SparkMax;

import static edu.wpi.first.units.Units.Amps;
import static frc.robot.constants.IntakeConstants.*;

public class IntakeIOReal implements IntakeIO {

    //  Roller 
    private final SparkMax rollerMotor = new SparkMax(INTAKE_ROLLER_MOTOR_ID, MotorType.kBrushless);
    private final RelativeEncoder rollerEncoder = rollerMotor.getEncoder();

    private final SimpleFF rollerFF = new SimpleFF(INTAKE_ROLLER_FF_GAINS);
    private final Alert rollerMotorDisconnect = new Alert("Intake roller Spark disconnected!", AlertType.kError);

    private boolean rollerOpenLoop = false;
    private AngularVelocity rollerGoal = Units.RPM.zero();

    //  Pivot 
    private final SparkMax pivotMotor = new SparkMax(INTAKE_PIVOT_MOTOR_ID, MotorType.kBrushless);
    private final CANcoder pivotEncoder = new CANcoder(INTAKE_PIVOT_ENCODER_ID);

    private final ProfiledPID pivotPID = new ProfiledPID(INTAKE_PIVOT_PID_GAINS);
    private ArmFF pivotFF = new ArmFF(INTAKE_PIVOT_FF_GAINS);

    private final Alert pivotMotorDisconnect = new Alert("Intake pivot Spark disconnected!", AlertType.kError);
    private final Alert pivotEncoderDisconnect = new Alert("Intake pivot encoder disconnected!", AlertType.kError);

    private boolean pivotOpenLoop = false;
    private Angle pivotGoal = INTAKE_STOWED_ANGLE;

    public IntakeIOReal() {
        // Live-tuning callbacks
        INTAKE_ROLLER_FF_GAINS.withCallback(() -> {
            rollerFF.setKs(INTAKE_ROLLER_FF_GAINS.kS);
            rollerFF.setKv(INTAKE_ROLLER_FF_GAINS.kV);
            rollerFF.setKa(INTAKE_ROLLER_FF_GAINS.kA);
        });
        INTAKE_PIVOT_PID_GAINS.withCallback(() -> pivotPID.setGains(INTAKE_PIVOT_PID_GAINS));
        INTAKE_PIVOT_FF_GAINS.withCallback(() -> {
            pivotFF.setKs(INTAKE_PIVOT_FF_GAINS.kS);
            pivotFF.setKg(INTAKE_PIVOT_FF_GAINS.kG);
            pivotFF.setKv(INTAKE_PIVOT_FF_GAINS.kV);
            pivotFF.setKa(INTAKE_PIVOT_FF_GAINS.kA);
        });

        pivotPID.setTolerance(INTAKE_PIVOT_TOLERANCE.in(Units.Radians));

        // Roller motor config
        SparkMaxConfig rollerCfg = new SparkMaxConfig();
        rollerCfg.idleMode(IdleMode.kBrake)
                 .inverted(INTAKE_ROLLER_INVERTED)
                 .smartCurrentLimit((int)INTAKE_ROLLER_CURRENT_LIM.in(Amps));
        rollerMotor.configure(rollerCfg, ResetMode.kResetSafeParameters, PersistMode.kPersistParameters);

        // Pivot motor config
        SparkMaxConfig pivotCfg = new SparkMaxConfig();
        pivotCfg.idleMode(IdleMode.kBrake)
                .inverted(INTAKE_PIVOT_INVERTED)
                .smartCurrentLimit((int)INTAKE_PIVOT_CURRENT_LIM.in(Amps));
        pivotMotor.configure(pivotCfg, ResetMode.kResetSafeParameters, PersistMode.kPersistParameters);
    }

    @Override
    public void updateInputs(IntakeIOInputs input) {
        //  Roller closed-loop 
        if (!rollerOpenLoop) {
            double ff = rollerFF.calculate(rollerGoal.in(Units.RPM));
            rollerMotor.setVoltage(Units.Volts.of(ff));
        }

        input.rollerVelocity = Units.RPM.of(rollerEncoder.getVelocity());
        input.rollerGoal = rollerGoal;

        input.rollerVoltageOut = Units.Volts.of(rollerMotor.getAppliedOutput() * rollerMotor.getBusVoltage());
        input.rollerCurrentOut = Units.Amps.of(rollerMotor.getOutputCurrent());
        input.rollerTemp = Units.Celsius.of(rollerMotor.getMotorTemperature());
        
        input.rollerConnected = rollerMotor.getLastError() != REVLibError.kCANDisconnected;
        input.rollerOpenLoop = rollerOpenLoop;
        rollerMotorDisconnect.set(!input.rollerConnected);

        //  Pivot closed-loop 
        if (!pivotOpenLoop) {
            double pid = pivotPID.calculate(pivotEncoder.getAbsolutePosition().getValue().in(Units.Radians), pivotGoal.in(Units.Radians));
            double ff = pivotFF.calculate(pivotPID.getSetpoint().position, pivotPID.getSetpoint().velocity);
            pivotMotor.setVoltage(Units.Volts.of(pid + ff));
        }
        input.pivotAngle = pivotEncoder.getAbsolutePosition().getValue();
        input.pivotVelocity = pivotEncoder.getVelocity().getValue();

        input.pivotGoal = pivotGoal;
        input.pivotSetpointPos = Units.Radians.of(pivotPID.getSetpoint().position);
        input.pivotSetpointVel = Units.RadiansPerSecond.of(pivotPID.getSetpoint().velocity);
        input.pivotAtSetpoint = pivotPID.atSetpoint();

        input.pivotVoltageOut = Units.Volts.of(pivotMotor.getAppliedOutput() * pivotMotor.getBusVoltage());
        input.pivotCurrentOut = Units.Amps.of(pivotMotor.getOutputCurrent());
        input.pivotTemp = Units.Celsius.of(pivotMotor.getMotorTemperature());
        
        input.pivotMotorConnected = pivotMotor.getLastError() != REVLibError.kCANDisconnected;
        input.pivotEncoderConnected = pivotEncoder.isConnected();
        input.pivotOpenLoop = pivotOpenLoop;
        pivotMotorDisconnect.set(!input.pivotMotorConnected);
        pivotEncoderDisconnect.set(!input.pivotEncoderConnected);

    }

    @Override
    public void setRollerVelocity(AngularVelocity velocity) {
        rollerOpenLoop = false;
        rollerGoal = velocity;
    }

    @Override
    public void setRollerVoltage(Voltage voltage) {
        rollerOpenLoop = true;
        rollerMotor.setVoltage(voltage);
    }

    @Override
    public void setPivotGoal(Angle goal) {
        pivotOpenLoop = false;
        pivotGoal = goal;
    }

    @Override
    public void setPivotVoltage(Voltage voltage) {
        pivotOpenLoop = true;
        pivotMotor.setVoltage(voltage);
    }

}

File: src/main/kotlin/frc/robot/subsystems/intake/IntakeIOReal.kt
package frc.robot.subsystems.intake

import com.ctre.phoenix6.hardware.CANcoder
import com.revrobotics.PersistMode
import com.revrobotics.REVLibError
import com.revrobotics.ResetMode
import com.revrobotics.spark.SparkLowLevel.MotorType
import com.revrobotics.spark.config.SparkBaseConfig.IdleMode
import com.revrobotics.spark.config.SparkMaxConfig
import edu.wpi.first.units.Units
import edu.wpi.first.units.Units.Amps
import edu.wpi.first.units.Units.Radians
import edu.wpi.first.units.measure.Angle
import edu.wpi.first.units.measure.Voltage
import edu.wpi.first.wpilibj.Alert
import frc.robot.constants.IntakeConstants.INTAKE_PIVOT_CURRENT_LIM
import frc.robot.constants.IntakeConstants.INTAKE_PIVOT_ENCODER_ID
import frc.robot.constants.IntakeConstants.INTAKE_PIVOT_FF_GAINS
import frc.robot.constants.IntakeConstants.INTAKE_PIVOT_INVERTED
import frc.robot.constants.IntakeConstants.INTAKE_PIVOT_MOTOR_ID
import frc.robot.constants.IntakeConstants.INTAKE_PIVOT_PID_GAINS
import frc.robot.constants.IntakeConstants.INTAKE_PIVOT_TOLERANCE
import frc.robot.constants.IntakeConstants.INTAKE_ROLLER_CURRENT_LIM
import frc.robot.constants.IntakeConstants.INTAKE_ROLLER_INVERTED
import frc.robot.constants.IntakeConstants.INTAKE_ROLLER_MOTOR_ID
import frc.robot.constants.IntakeConstants.INTAKE_STOWED_ANGLE
import frc.utils.controlWrappers.ArmFF
import frc.utils.controlWrappers.ProfiledPID
import frc.utils.motorWrappers.SparkMax

class IntakeIOReal : IntakeIO {

    // Roller
    private val rollerMotor = SparkMax(INTAKE_ROLLER_MOTOR_ID, MotorType.kBrushless)
    private val rollerEncoder = rollerMotor.getEncoder()

    private val rollerMotorDisconnect = Alert("Intake roller Spark disconnected!", Alert.AlertType.kError)

    // Pivot
    private val pivotMotor = SparkMax(INTAKE_PIVOT_MOTOR_ID, MotorType.kBrushless)
    private val pivotEncoder = CANcoder(INTAKE_PIVOT_ENCODER_ID)

    private val pivotPID = ProfiledPID(INTAKE_PIVOT_PID_GAINS)
    private var pivotFF = ArmFF(INTAKE_PIVOT_FF_GAINS)

    private val pivotMotorDisconnect = Alert("Intake pivot Spark disconnected!", Alert.AlertType.kError)
    private val pivotEncoderDisconnect = Alert("Intake pivot encoder disconnected!", Alert.AlertType.kError)

    private var pivotOpenLoop = false
    private var pivotGoal: Angle = INTAKE_STOWED_ANGLE

    init {
        // Live-tuning callbacks
        INTAKE_PIVOT_PID_GAINS.withCallback { pivotPID.setGains(INTAKE_PIVOT_PID_GAINS) }
        INTAKE_PIVOT_FF_GAINS.withCallback {
            pivotFF.setKs(INTAKE_PIVOT_FF_GAINS.kS)
            pivotFF.setKg(INTAKE_PIVOT_FF_GAINS.kG)
            pivotFF.setKv(INTAKE_PIVOT_FF_GAINS.kV)
            pivotFF.setKa(INTAKE_PIVOT_FF_GAINS.kA)
        }

        pivotPID.setTolerance(INTAKE_PIVOT_TOLERANCE.`in`(Units.Radians))

        // Roller motor config
        val rollerCfg = SparkMaxConfig()
        rollerCfg.idleMode(IdleMode.kBrake)
            .inverted(INTAKE_ROLLER_INVERTED)
            .smartCurrentLimit(INTAKE_ROLLER_CURRENT_LIM.`in`(Amps).toInt())
        rollerMotor.configure(rollerCfg, ResetMode.kResetSafeParameters, PersistMode.kPersistParameters)

        // Pivot motor config
        val pivotCfg = SparkMaxConfig()
        pivotCfg.idleMode(IdleMode.kBrake)
            .inverted(INTAKE_PIVOT_INVERTED)
            .smartCurrentLimit(INTAKE_PIVOT_CURRENT_LIM.`in`(Amps).toInt())
        pivotMotor.configure(pivotCfg, ResetMode.kResetSafeParameters, PersistMode.kPersistParameters)
    }

    override fun updateInputs(input: IntakeIO.IntakeIOInputs) {
        // Roller closed-loop
        input.rollerVelocity = Units.RPM.of(rollerEncoder.velocity)

        input.rollerVoltageOut = Units.Volts.of(rollerMotor.appliedOutput * rollerMotor.busVoltage)
        input.rollerCurrentOut = Units.Amps.of(rollerMotor.outputCurrent)
        input.rollerTemp = Units.Celsius.of(rollerMotor.motorTemperature)

        input.rollerConnected = rollerMotor.getLastError() != REVLibError.kCANDisconnected
        rollerMotorDisconnect.set(!input.rollerConnected)

        // Pivot closed-loop
        if (!pivotOpenLoop) {
            val pid = pivotPID.calculate(pivotEncoder.absolutePosition.value.`in`(Units.Radians), pivotGoal.`in`(Units.Radians))
            val ff = pivotFF.calculate(pivotEncoder.absolutePosition.value.`in`(Radians), pivotPID.setpoint.velocity)
            pivotMotor.setVoltage(Units.Volts.of(pid + ff))
        }
        input.pivotAngle = pivotEncoder.absolutePosition.value
        input.pivotVelocity = pivotEncoder.velocity.value

        input.pivotGoal = pivotGoal
        input.pivotSetpointPos = Units.Radians.of(pivotPID.setpoint.position)
        input.pivotSetpointVel = Units.RadiansPerSecond.of(pivotPID.setpoint.velocity)
        input.pivotAtSetpoint = pivotPID.atSetpoint()

        input.pivotVoltageOut = Units.Volts.of(pivotMotor.appliedOutput * pivotMotor.busVoltage)
        input.pivotCurrentOut = Units.Amps.of(pivotMotor.outputCurrent)
        input.pivotTemp = Units.Celsius.of(pivotMotor.motorTemperature)

        input.pivotMotorConnected = pivotMotor.getLastError() != REVLibError.kCANDisconnected
        input.pivotEncoderConnected = pivotEncoder.isConnected()
        input.pivotOpenLoop = pivotOpenLoop
        pivotMotorDisconnect.set(!input.pivotMotorConnected)
        pivotEncoderDisconnect.set(!input.pivotEncoderConnected)
    }

    override fun setRollerVoltage(voltage: Voltage) {
        rollerMotor.setVoltage(voltage)
    }

    override fun setPivotGoal(goal: Angle) {
        pivotOpenLoop = false
        pivotGoal = goal
    }

    override fun setPivotVoltage(voltage: Voltage) {
        pivotOpenLoop = true
        pivotMotor.setVoltage(voltage)
    }
}
